composer/utils/checkpoint.py

```diff
@@ -200,46 +201,234 @@ def load_checkpoint(
+    state: State,
+    logger: Logger,
+    object_store: Optional[Union[ObjectStore, LoggerDestination]] = None,
+    load_weights_only: bool = False,
+    strict_model_weights: bool = False,
+    progress_bar: bool = True,
+    ignore_keys: Optional[Union[List[str], Callable[[Dict], None]]] = None,
+    exclude_algorithms: Optional[List[str]] = None,
+    algorithm_passes: Optional[List[AlgorithmPass]] = None,
+) -> Dict:
+
+    if not using_torch_2():
+        raise ValueError(f'Sharded checkpoint loading requires torch version >= 2.0.0. You have torch version {torch.__version__}')
+
+    from torch.distributed import checkpoint as dist_cp
+    from torch.distributed.checkpoint.metadata import Metadata
+    from torch.distributed.checkpoint.optimizer import load_sharded_optimizer_state_dict
+    from torch.distributed.checkpoint.planner import LoadPlan, LoadPlanner
+
+    # This function is used so we can figure out which ranks need to load saved rngs and which can just make their own.
+    def _get_num_ranks_that_saved_rng(metadata: Metadata):
+        rng_inds = []
+        for field_name, field_value in metadata.planner_data.items():
+            if 'rng' in field_name:
+                _, rng_rank_index, _ = field_value
+                rng_inds.append(rng_rank_index)
+        rng_inds = set(rng_inds)
+        return len(rng_inds)
+
+    # A subclass of FileSystemReader that downloads files from the object store before reading them from the local filesystem.
+    class DistCPObjectStoreReader(dist_cp.FileSystemReader):
+
+        def __init__(self, source_path: str, destination_path: str, object_store):  # path to metadata
+            self.source_path = source_path
+            self.destination_path = destination_path
+            self.object_store = object_store
+
+            # Download metadata file.
+            Path(self.destination_path).mkdir(parents=True, exist_ok=True)
+            metadata_destination = os.path.join(self.destination_path, '.metadata')
+            if dist.get_local_rank() == 0:
+                object_store.download_object(object_name=str(Path(source_path) / Path('.metadata')),
+                                             filename=metadata_destination)
+            dist.barrier()
+                
+            # Instantiate FileSystemReader with destination path b/c that's where we will download files to.
+            # Because we already download the metadata file to the destination, everything will work swimmingly
+            super().__init__(destination_path)
+
+        def read_data(self, plan: LoadPlan, planner: LoadPlanner):
+            # 1. Download to the destination all files that this rank is responsible for.
+            for plan_item in plan.items:
+                relative_file_path = self.storage_data[plan_item.storage_index].relative_path
+                file_destination = str(Path(self.destination_path) / Path(relative_file_path))
+                if not os.path.exists(file_destination):
```

Comment:
the reason for the check is the same file can show up multiple times in the plan and even multiple ranks can see the same file, so we're just avoiding downloading the same file multiple times. there is not risk for overwriting because this is a temp file

Reply:
added a comment to explain

Reply:
ah, didn't know that, thanks